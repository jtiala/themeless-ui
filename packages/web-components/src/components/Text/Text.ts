/* eslint-disable lit/no-invalid-html */
/* eslint-disable lit/binding-positions */
import style from "@themeless-ui/style/dist/text.css?inline";
import { cn } from "@themeless-ui/utils";
import { nothing, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { html, unsafeStatic } from "lit/static-html.js";
import { TUIComponent } from "../TUIComponent";

type TextType =
  | "abbr"
  | "b"
  | "cite"
  | "code"
  | "del"
  | "em"
  | "i"
  | "ins"
  | "kbd"
  | "mark"
  | "q"
  | "s"
  | "samp"
  | "small"
  | "span"
  | "strong"
  | "sub"
  | "sup"
  | "u"
  | "var";

const className = cn("text");

@customElement("tui-text")
export class Text extends TUIComponent {
  /**
   * The type of the rendered element.
   *
   * @default span
   */
  @property()
  type!: TextType;

  /**
   * Only available when `type` is `abbr`.
   *
   * It's recommended to provide a full expansion of the abbreviated term in plain text on the first use of the abbreviation.
   * If it's not viable to provide the expansion as full text, `title` should be used instead.
   */
  @property()
  abbrTitle?: string;

  /**
   * Only available when `type` is `q`, `del`, or `ins`.
   *
   * `q`: A URI of the source of the information quoted.
   * `del` or `ins`: A URI for a resource that explains the change.
   */
  @property()
  cite?: string;

  /**
   * Only available when `type` is `del` or `ins`.
   *
   * Indicates the time and date of the change. Must be a valid date string with an optional time.
   */
  @property()
  dateTime?: string;

  static styles = unsafeCSS(style);

  constructor() {
    super();

    this.type = "span";
  }

  override render() {
    const tag = unsafeStatic(this.type);

    return html`
      <${tag}
        title="${this.abbrTitle || nothing}"
        cite="${this.cite || nothing}"
        dateTime="${this.dateTime || nothing}"
        class="${className}"
        id="${this.id || nothing}"
        data-testid="${this.testId || nothing}"
      >
        <slot></slot>
      </${tag}>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-text": Text;
  }
}
