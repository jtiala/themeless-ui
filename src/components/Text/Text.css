.tui__text {
  display: inline;
}
