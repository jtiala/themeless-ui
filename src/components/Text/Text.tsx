import { createElement, ReactNode } from "react";
import { CommonComponentProps } from "../../utils/components";
import { cn } from "../../utils/css";

type TextType =
  | "abbr"
  | "b"
  | "cite"
  | "code"
  | "del"
  | "em"
  | "i"
  | "ins"
  | "kbd"
  | "mark"
  | "q"
  | "s"
  | "samp"
  | "small"
  | "span"
  | "strong"
  | "sub"
  | "sup"
  | "u"
  | "var";

type GeneralProps = {
  /**
   * The type of the rendered element.
   *
   * @default span
   */
  type?: Extract<
    TextType,
    | "b"
    | "cite"
    | "code"
    | "del"
    | "em"
    | "i"
    | "ins"
    | "kbd"
    | "mark"
    | "s"
    | "samp"
    | "small"
    | "span"
    | "strong"
    | "sub"
    | "sup"
    | "u"
    | "var"
  >;

  /**
   * Only available when `type` is `abbr`.
   *
   * It's recommended to provide a full expansion of the abbreviated term in plain text on the first use of the abbreviation.
   * If it's not viable to provide the expansion as full text, `title` should be used instead.
   */
  title?: never;

  /**
   * Only available when `type` is `q`, `del`, or `ins`.
   *
   * `q`: A URI of the source of the information quoted.
   * `del` or `ins`: A URI for a resource that explains the change.
   */
  cite?: never;

  /**
   * Only available when `type` is `del` or `ins`.
   *
   * Indicates the time and date of the change. Must be a valid date string with an optional time.
   */
  dateTime?: never;
};

type AbbreviationProps = {
  type: Extract<TextType, "abbr">;
  title?: string;
  cite?: never;
  dateTime?: never;
};

type QuoteProps = {
  type: Extract<TextType, "q">;
  title?: never;
  cite?: string;
  dateTime?: never;
};

type DeletedTextProps = {
  type: Extract<TextType, "del">;
  title?: never;
  cite?: string;
  dateTime?: string;
};

type InsertedTextProps = {
  type: Extract<TextType, "ins">;
  title?: never;
  cite?: string;
  dateTime?: string;
};

type CommonProps = {
  /**
   * Text content.
   */
  children?: ReactNode;
};

export type TextProps = (
  | GeneralProps
  | AbbreviationProps
  | DeletedTextProps
  | InsertedTextProps
  | QuoteProps
) &
  CommonProps &
  CommonComponentProps;

const className = cn("text");

/**
 * Display inline text content with different styles.
 */
export function Text({ type, children, testId, ...rest }: TextProps) {
  return createElement(
    type || "span",
    { className, ...rest, "data-testid": testId },
    children
  );
}
