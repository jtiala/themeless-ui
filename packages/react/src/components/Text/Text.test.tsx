import {
  cleanup,
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Text, TextProps } from "./Text";

const defaultText = <Text testId="text">Lorem ipsum dolor sit amet.</Text>;

describe("Text", async () => {
  it("should render the text", () => {
    expect(componentToJson(renderer.create(defaultText))).toMatchSnapshot();

    render(defaultText);

    const text = screen.getByTestId("text");

    expect(text).toBeInTheDocument();
    expect(text).toHaveTextContent("Lorem ipsum dolor sit amet.");
    expect(text.tagName.toLowerCase()).toBe("span");
  });

  it("should render correct html element", () => {
    const types: NonNullable<TextProps["type"]>[] = [
      "abbr",
      "b",
      "cite",
      "code",
      "del",
      "em",
      "i",
      "ins",
      "kbd",
      "mark",
      "q",
      "s",
      "samp",
      "small",
      "span",
      "strong",
      "sub",
      "sup",
      "u",
      "var",
    ];

    types.forEach((type) => {
      cleanup();

      const element = (
        <Text type={type} testId="text">
          {type}
        </Text>
      );

      expect(componentToJson(renderer.create(element))).toMatchSnapshot();

      render(element);

      const text = screen.getByTestId("text");

      expect(text).toBeInTheDocument();
      expect(text).toHaveTextContent(type);
      expect(text.tagName.toLowerCase()).toBe(type);
    });
  });
});
