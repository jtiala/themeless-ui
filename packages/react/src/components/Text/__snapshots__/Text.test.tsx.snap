// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Text > should render correct html element 1`] = `
<abbr
  className="tui__text"
  data-testid="text"
>
  abbr
</abbr>
`;

exports[`Text > should render correct html element 2`] = `
<b
  className="tui__text"
  data-testid="text"
>
  b
</b>
`;

exports[`Text > should render correct html element 3`] = `
<cite
  className="tui__text"
  data-testid="text"
>
  cite
</cite>
`;

exports[`Text > should render correct html element 4`] = `
<code
  className="tui__text"
  data-testid="text"
>
  code
</code>
`;

exports[`Text > should render correct html element 5`] = `
<del
  className="tui__text"
  data-testid="text"
>
  del
</del>
`;

exports[`Text > should render correct html element 6`] = `
<em
  className="tui__text"
  data-testid="text"
>
  em
</em>
`;

exports[`Text > should render correct html element 7`] = `
<i
  className="tui__text"
  data-testid="text"
>
  i
</i>
`;

exports[`Text > should render correct html element 8`] = `
<ins
  className="tui__text"
  data-testid="text"
>
  ins
</ins>
`;

exports[`Text > should render correct html element 9`] = `
<kbd
  className="tui__text"
  data-testid="text"
>
  kbd
</kbd>
`;

exports[`Text > should render correct html element 10`] = `
<mark
  className="tui__text"
  data-testid="text"
>
  mark
</mark>
`;

exports[`Text > should render correct html element 11`] = `
<q
  className="tui__text"
  data-testid="text"
>
  q
</q>
`;

exports[`Text > should render correct html element 12`] = `
<s
  className="tui__text"
  data-testid="text"
>
  s
</s>
`;

exports[`Text > should render correct html element 13`] = `
<samp
  className="tui__text"
  data-testid="text"
>
  samp
</samp>
`;

exports[`Text > should render correct html element 14`] = `
<small
  className="tui__text"
  data-testid="text"
>
  small
</small>
`;

exports[`Text > should render correct html element 15`] = `
<span
  className="tui__text"
  data-testid="text"
>
  span
</span>
`;

exports[`Text > should render correct html element 16`] = `
<strong
  className="tui__text"
  data-testid="text"
>
  strong
</strong>
`;

exports[`Text > should render correct html element 17`] = `
<sub
  className="tui__text"
  data-testid="text"
>
  sub
</sub>
`;

exports[`Text > should render correct html element 18`] = `
<sup
  className="tui__text"
  data-testid="text"
>
  sup
</sup>
`;

exports[`Text > should render correct html element 19`] = `
<u
  className="tui__text"
  data-testid="text"
>
  u
</u>
`;

exports[`Text > should render correct html element 20`] = `
<var
  className="tui__text"
  data-testid="text"
>
  var
</var>
`;

exports[`Text > should render the text 1`] = `
<span
  className="tui__text"
  data-testid="text"
>
  Lorem ipsum dolor sit amet.
</span>
`;
